File: src/player.cpp
#include "player.hpp"
#include "settings.hpp"
#include "support.hpp"
#include "upgrade.hpp"
#include <SFML/System/Time.hpp>
#include <SFML/System/Vector2.hpp>
#include <SFML/Window/Keyboard.hpp>
#include <SFML/Graphics/Rect.hpp>
#include <algorithm>
#include <experimental/filesystem>

#include <cassert>
#include <memory>
#include <sstream>
#include <vector>
#include <functional>

Player::Player(const sf::Vector2f &pos,
               const std::list<std::shared_ptr<SpriteTexture>> &obstacle_sprites,
               const std::function<void()> &create_attack,
               const std::function<void()> &destroy_attack,
               const std::function<void(const std::string &, int, int)> &create_magic,
               const std::function<void()> &destroy_magic,
               const std::function<void(int, const std::string &)> &damage_player)
    : Entity(0, 0.15f, {0, 0}, obstacle_sprites),
      status_("down"),
      speed(5),
      attacking(false),
      attack_cooldown(sf::milliseconds(400)),
      weapon_index_(0),
      magic_index_(0),
      can_switch_weapon_(true),
      switch_duration_cooldown(sf::milliseconds(200)),
      can_switch_magic_(true),
      vulernable_(true),
      invincibility_duration(sf::milliseconds(300)),
      create_attack(create_attack),
      destroy_attack(destroy_attack),
      create_magic(create_magic),
      destroy_magic(destroy_magic),
      damage_player_(damage_player)
{
  auto r = this->loadFromFile("./src/graphics/player.png");
  assert(r);
  this->get_rect({"topleft", pos});
  this->hitbox_ = this->rect_.inflate(0, -26);

  this->weapon_ = nth_name(WEAPON_DATA, this->weapon_index_);
  this->magic = nth_name(MAGIC_DATA, this->magic_index_);

  this->import_player_assets();

  this->stats_ = PLAYER_STATS;
  this->max_stats_ = PLAYER_MAX_STATS;
  this->upgrade_cost_ = UPGRADE_COST;
  this->health_ = this->stats_["health"];
  this->energy_ = this->stats_["energy"];
  this->exp_ = 0;

  this->weapon_attacks_sound_buffer.loadFromFile("./src/audio/sword.wav");
  this->weapon_attacks_sound.setBuffer(this->weapon_attacks_sound_buffer);
  this->weapon_attacks_sound.setVolume(10);
}

void Player::import_player_assets()
{
  auto animation_names = {
      "down", "down_attack", "down_idle", "left", "left_attack", "left_idle",
      "right", "right_attack", "right_idle", "up", "up_attack", "up_idle"};
  auto base_path = std::experimental::filesystem::path("./src/graphics/player");
  for (const auto &name : animation_names)
  {
    this->animations[name] = import_folder((base_path / name).string());
  }
}

void Player::input()
{
  if (this->attacking)
    return;

  // movement
  if (sf::Keyboard::isKeyPressed(sf::Keyboard::Left))
  {
    this->direction.x = -1;
    this->status_ = "left";
  }
  else if (sf::Keyboard::isKeyPressed(sf::Keyboard::Right))
  {
    this->direction.x = 1;
    this->status_ = "right";
  }
  else
    this->direction.x = 0;

  if (sf::Keyboard::isKeyPressed(sf::Keyboard::Up))
  {
    this->direction.y = -1;
    this->status_ = "up";
  }
  else if (sf::Keyboard::isKeyPressed(sf::Keyboard::Down))
  {
    this->direction.y = 1;
    this->status_ = "down";
  }
  else
    this->direction.y = 0;

  // attack
  if (sf::Keyboard::isKeyPressed(sf::Keyboard::Z))
  {
    this->attacking = true;
    this->attack_time.restart();
    this->create_attack();
    this->weapon_attacks_sound.play();
  }

  // magic
  if (sf::Keyboard::isKeyPressed(sf::Keyboard::X))
  {
    this->attacking = true;
    this->attack_time.restart();
    auto magic_data = MAGIC_DATA[this->magic];
    this->create_magic(this->magic,
                       magic_data.strength,
                       magic_data.cost);
  }

  // swap weapon
  if (this->can_switch_weapon_ &&
      sf::Keyboard::isKeyPressed(sf::Keyboard::Tab))
  {
    this->can_switch_weapon_ = false;
    this->weapon_switch_time.restart();
    this->weapon_index_++;
    if (this->weapon_index_ >= WEAPON_DATA.size())
    {
      this->weapon_index_ = 0;
    }
    this->weapon_ = nth_name<WeaponData>(WEAPON_DATA, this->weapon_index_);
  }

  // swap magic::Z
  if (this->can_switch_magic_ && sf::Keyboard::isKeyPressed(sf::Keyboard::Q))
  {
    this->can_switch_magic_ = false;
    this->magic_switch_time.restart();
    this->magic_index_++;
    if (this->magic_index_ >= MAGIC_DATA.size())
    {
      this->magic_index_ = 0;
    }
    this->magic = nth_name(MAGIC_DATA, this->magic_index_);
  }
}

void Player::cooldowns()
{
  if (this->attacking)
  {
    if (this->attack_time.getElapsedTime() > (this->attack_cooldown + sf::milliseconds(WEAPON_DATA[this->weapon_].cooldown)))
    {
      this->attacking = false;
      this->destroy_attack();
      //?? destroy_magic 은 언제 어디서 호출되어야 할까?
    }
  }

  if (!this->can_switch_weapon_)
  {
    if (this->weapon_switch_time.getElapsedTime() > this->switch_duration_cooldown)
    {
      this->can_switch_weapon_ = true;
    }
  }

  if (!this->can_switch_magic_)
  {
    if (this->magic_switch_time.getElapsedTime() > this->switch_duration_cooldown)
    {
      this->can_switch_magic_ = true;
    }
  }

  if (!this->vulernable_)
  {
    if (this->hurt_time.getElapsedTime() > this->invincibility_duration)
    {
      this->vulernable_ = true;
    }
  }
}

void Player::get_status()
{
  if (this->direction.x == 0 && this->direction.y == 0)
  {
    if (std::string::npos == this->status_.find("idle") &&
        std::string::npos == this->status_.find("attack"))
    { // not idle or attack
      std::stringstream ss;
      ss << this->status_ << "_idle";
      this->status_ = ss.str();
    }
  }

  if (this->attacking)
  {
    this->direction.x = 0;
    this->direction.y = 0;
    if (std::string::npos == this->status_.find("attack"))
    {
      if (std::string::npos != this->status_.find("idle"))
      { // found
        auto index = this->status_.find("idle");
        this->status_ = this->status_.substr(0, index) + "attack";
      }
      else
      {
        this->status_ = this->status_ + "_attack";
      }
    }
  }
  else
  {
    if (std::string::npos != this->status_.find("attack"))
    { // found
      auto index = this->status_.find("attack");
      this->status_ = this->status_.substr(0, index) + "idle";
    }
  }
}

void Player::update_sprite(std::shared_ptr<SpriteTexture> sprite)
{
  this->texture_ = sprite->texture_; // 이걸 public 이 아닌 protected 로 바꿀수가 없을까?
  this->sprite_.setTexture(*this->texture_);
  this->rect_ = sprite->rect();
}

void Player::animate()
{
  auto animation = this->animations[this->status_];
  this->frame_index += this->animation_speed;
  if (this->frame_index >= static_cast<float>(animation.size()))
  {
    this->frame_index = 0;
  }

  this->update_sprite(animation[static_cast<int>(this->frame_index)]);
  this->get_rect({"center", this->hitbox_.center()});

  if (!this->vulernable_)
  { // make flicker
    auto alpha = this->wave_value();
    this->sprite_.setColor(sf::Color(0xff, 0xff, 0xff, alpha));
  }
  else
  {
    this->sprite_.setColor(sf::Color(0xff, 0xff, 0xff, 0xff));
  }
}
void Player::update()
{
  this->input();
  this->cooldowns();
  this->get_status();
  this->animate();
  this->move(this->stats_["speed"]);
  this->energy_recovery();
}

int Player::get_full_weapon_damage()
{
  auto base_damage = this->stats_["attack"];
  auto weapon_damage = WEAPON_DATA[this->weapon_].damage;
  return base_damage + weapon_damage;
}

int Player::get_full_magic_damage()
{
  auto base_damage = this->stats_["attack"];
  auto magic_damage = MAGIC_DATA[this->magic].strength;
  return base_damage + magic_damage;
}

void Player::energy_recovery()
{
  if (this->energy_ <= this->stats_["energy"])
  {
    this->energy_ += 0.01 * this->stats_["magic"];
  }
  else
  {
    this->energy_ = this->stats_["energy"];
  }
}

File: pyZelda/player.py
import pygame
from entity import Entity

from support import import_folder
from settings import HITBOX_OFFSET, weapon_data
from settings import magic_data

class Player(Entity):
    def __init__(self, pos, groups, obstacle_sprites, 
            create_attack, destroy_attack,
            create_magic, destroy_magic, damage_player) -> None:
        super().__init__(groups)
        self.image = pygame.image.load("./graphics/player.png").convert_alpha()
        self.rect = self.image.get_rect(topleft=pos)
        self.hitbox = self.rect.inflate(-6, HITBOX_OFFSET['player'])

        self.import_player_assets()
        self.status = 'down'
        # self.frame_index = 0
        # self.animation_speed = 0.15
       
        # self.direction = pygame.math.Vector2()
        self.attacking = False
        self.attack_cooldown = 400
        self.attack_time = None

        self.obstacles_sprites = obstacle_sprites

        # weapon
        self.create_attack = create_attack
        self.destroy_attack = destroy_attack
        self.weapon_index = 0
        self.weapon = list(weapon_data.keys())[self.weapon_index]

        self.can_switch_weapon = True
        self.weapon_switch_time = None
        self.switch_duration_cooldown = 200

        # magic
        self.create_magic = create_magic
        self.destroy_magic = destroy_magic
        self.magic_index = 0
        self.magic = list(magic_data.keys())[self.magic_index]

        self.can_switch_magic = True
        self.magic_switch_time = None


        # stats
        self.stats = { 'health': 100, 'energy':60, 'attack': 10, 'magic':4, 'speed':6}
        self.max_stats = { 'health': 300, 'energy':140, 'attack': 20, 'magic':10, 'speed':12}
        self.upgrade_cost = { 'health': 100, 'energy':100, 'attack': 100, 'magic':100, 'speed':100}
        self.health = self.stats['health']
        self.energy = self.stats['energy']
        self.exp = 0 # 경험치


        self.damage_player = damage_player

        # invincibility timer
        self.vulernable = True
        self.hurt_time = None
        self.invincibility_duration = 300

        # sound
        self.weapon_attacks_sound = pygame.mixer.Sound("./audio/sword.wav")
        self.weapon_attacks_sound.set_volume(0.1)

    def get_value_by_index(self, index):
        return list(self.stats.values())[index]

    def get_cost_by_index(self, index):
        return list(self.upgrade_cost.values())[index]

    def energy_recovery(self):
        if self.energy <= self.stats['energy']:
            self.energy += 0.01 * self.stats['magic']
        else:
            self.energy = self.stats['energy']

    def get_full_weapon_damage(self):
        base_damage = self.stats['attack']
        weapon_damage = weapon_data[self.weapon]['damage']
        return base_damage + weapon_damage

    def get_full_magic_damage(self):
        base_damage = self.stats['attack']
        magic_damage = magic_data[self.magic]['strength']
        return base_damage + magic_damage

    def import_player_assets(self):
        self.animations = {
            "down":[],
            "down_attack":[],
            "down_idle": [],
            "left": [],
            "left_attack": [],
            "left_idle": [],
            "right": [],
            "right_attack": [],
            "right_idle": [],
            "up": [],
            "up_attack": [],
            "up_idle": []
        }

        for animation in self.animations.keys():
            self.animations[animation] = import_folder(f"./graphics/player/{animation}")

    def input(self) -> None:
        if self.attacking: return None
        
        keys = pygame.key.get_pressed()

        # movement
        if keys[pygame.K_LEFT]:
            self.direction.x = -1
            self.status = "left"
        elif keys[pygame.K_RIGHT]:
            self.direction.x = 1
            self.status = "right"
        else:
            self.direction.x = 0

        if keys[pygame.K_UP]:
            self.direction.y = -1
            self.status = "up"
        elif keys[pygame.K_DOWN]:
            self.direction.y = 1
            self.status = "down"
        else:
            self.direction.y = 0

        # attack
        if keys[pygame.K_z]:
            self.attacking = True
            self.attack_time = pygame.time.get_ticks()
            self.create_attack()
            self.weapon_attacks_sound.play()

        # magic
        if keys[pygame.K_x]:
            self.attacking = True
            self.attack_time = pygame.time.get_ticks()
            # style = magic_data[self.magic]['graphic']
            style = self.magic
            strength = magic_data[self.magic]['strength']
            cost = magic_data[self.magic]['cost']
            self.create_magic(style, strength, cost)

        # swap weapon
        if keys[pygame.K_TAB] and self.can_switch_weapon:
            self.can_switch_weapon = False
            self.weapon_switch_time = pygame.time.get_ticks()
            self.weapon_index += 1
            if self.weapon_index >= len(weapon_data):
                self.weapon_index = 0
            self.weapon = list(weapon_data.keys())[self.weapon_index]

        # swap magic
        if keys[pygame.K_q] and self.can_switch_magic:
            self.can_switch_magic = False
            self.magic_switch_time = pygame.time.get_ticks()
            self.magic_index += 1
            if self.magic_index >= len(magic_data):
                self.magic_index = 0
            self.magic = list(magic_data.keys())[self.magic_index]


    def get_status(self):
        if self.direction.x == 0 and self.direction.y == 0:
            if not 'idle' in self.status and not 'attack' in self.status:
                self.status = f"{self.status}_idle"

        if self.attacking:
            self.direction.x = 0
            self.direction.y = 0
            if not 'attack' in self.status:
                if 'idle' in self.status:
                    self.status = self.status.replace('idle', 'attack')
                else:
                    self.status = f"{self.status}_attack"
        else:
            if 'attack' in self.status:
                self.status = self.status.replace('attack', 'idle')

    # def collision(self, direction):
    #     if direction == "horizontal":
    #         for sprite in self.obstacles_sprites:
    #             if sprite.hitbox.colliderect(self.hitbox):
    #                 if self.direction.x > 0:  # moving right
    #                     self.hitbox.right = sprite.hitbox.left
    #                 elif self.direction.x < 0:  # moving left
    #                     self.hitbox.left = sprite.hitbox.right
    #     elif direction == "vertical":
    #         for sprite in self.obstacles_sprites:
    #             if sprite.hitbox.colliderect(self.hitbox):
    #                 if self.direction.y > 0:  # moving down
    #                     self.hitbox.bottom = sprite.hitbox.top
    #                 elif self.direction.y < 0:  # moving up
    #                     self.hitbox.top = sprite.hitbox.bottom

    # def move(self, speed):
    #     # make same speed in x and y direction
    #     if self.direction.magnitude() != 0:
    #         self.direction = self.direction.normalize()

    #     self.hitbox.x += self.direction.x * speed
    #     self.collision("horizontal")
    #     self.hitbox.y += self.direction.y * speed
    #     self.collision("vertical")

    #     self.rect.center = self.hitbox.center

    def cooldowns(self):
        current_time = pygame.time.get_ticks()
        if self.attacking:
            if current_time - self.attack_time > self.attack_cooldown + weapon_data[self.weapon]['cooldown']:
                self.attacking = False
                self.destroy_attack()
                
        if not self.can_switch_weapon:
            if current_time - self.weapon_switch_time > self.switch_duration_cooldown:
                self.can_switch_weapon = True

        if not self.can_switch_magic:
            if current_time - self.magic_switch_time > self.switch_duration_cooldown:
                self.can_switch_magic = True

        if not self.vulernable:
            if current_time - self.hurt_time > self.invincibility_duration:
                self.vulernable = True

    def animate(self):
        animation = self.animations[self.status]

        self.frame_index += self.animation_speed
        if self.frame_index >= len(animation):
            self.frame_index = 0

        self.image = animation[int(self.frame_index)]
        self.rect = self.image.get_rect(center=self.hitbox.center)

        if not self.vulernable:
            # flicker
            alpha = self.wave_value()
            self.image.set_alpha(alpha)
        else:
            self.image.set_alpha(255)


    def update(self) -> None:
        self.input()
        self.cooldowns()
        self.get_status()
        self.animate()
        self.move(self.stats['speed'])
        self.energy_recovery()
